accept md and rmd files; refactor align, shrink, and calc_column_stats to work on address/visual range

carywreams opened the pull request:
+ plugin/rainbow_csv.vim converts RainbowAlign and RainbowShrink to use ``-range=%``
+ autoload/rainbow_csv.vim#csv_align
  + accepts either address or visual range
  + changes operational loops to have specific start/stop  lines
  + emits error if an entire markdown or rmd file is submitted
+ autoload/rainbow_csv.vim#csv_shrink
  + accepts either address or visual range
  + changes operational loops to have specific start/stop  lines
  + emits error if an entire markdown or rmd file is submitted
+ autoload/rainbow_csv.vim#calc_column_stats
  + accepts two new parameters, firstline and lastline
  + changes operational loops to have specific start/stop  lines
+ autoload/rainbow_csv.vim#s:named_syntax_map adds markdown and rmd filetype entries
+ proposed README.md changes
+ proposed doc/rainbow_csv.txt changes

carywreams committed:
issue #51: initial conversion of csv_align to accept range of lines

carywreams committed:
issue #51: range-enabled csv_shrink; converts csv_shrink, csv_align to use either address or visual range

carywreams committed:
issue #51: adds documentation of md and rmd filetype acceptance as well as requirements for ranges when using align and shrink functions in those filetypes

carywreams commented:
addresses #51

mechatroner commented:
Awesome! Thank you! Looking at it

mechatroner commented on autoload/rainbow_csv.vim:
```diff
@@ -17,7 +17,7 @@ let s:system_python_interpreter = ''
 let s:magic_chars = '^*$.~/[]\'
 
-let s:named_syntax_map = {'csv': [',', 'quoted', ''], 'csv_semicolon': [';', 'quoted', ''], 'tsv': ["\t", 'simple', ''], 'csv_pipe': ['|', 'simple', ''], 'csv_whitespace': [" ", 'whitespace', ''], 'rfc_csv': [',', 'quoted_rfc', ''], 'rfc_semicolon': [';', 'quoted_rfc', '']}
+let s:named_syntax_map = {'csv': [',', 'quoted', ''], 'csv_semicolon': [';', 'quoted', ''], 'tsv': ["\t", 'simple', ''], 'csv_pipe': ['|', 'simple', ''], 'csv_whitespace': [" ", 'whitespace', ''], 'rfc_csv': [',', 'quoted_rfc', ''], 'rfc_semicolon': [';', 'quoted_rfc', ''], 'markdown':['|', 'simple', ''], 'rmd':['|', 'simple', '']}
```
Please don't change this list since it can be used by other logic.

carywreams replied:
done

mechatroner commented on autoload/rainbow_csv.vim:
```diff
@@ -820,12 +818,13 @@ func! rainbow_csv#align_field(field, is_first_line, max_field_components_lens, i
     " Numeric-aware logic by itself adds about 50% runtime compared to the basic string-based field width alignment
     " If there are lot of numeric columns this can additionally increase runtime by another 50% or more.
-    let show_progress_bar = wordcount()['bytes'] > 200000
+    let show_progress_bar = (l:lastline - l:firstline) > 200000
```
Please revert, you replaced byte count with line count

carywreams replied:
done

mechatroner commented on autoload/rainbow_csv.vim:
```diff
@@ -820,12 +818,13 @@ func! rainbow_csv#align_field(field, is_first_line, max_field_components_lens, i
+    let show_progress_bar = (l:lastline - l:firstline) > 200000
     let [delim, policy, comment_prefix] = rainbow_csv#get_current_dialect()
     if policy == 'monocolumn'
         echoerr "RainbowAlign is available only for highlighted CSV files"
```
Instead of immediately returning you can adjust delim and policy to '|' and simple for markdown files - this would be the special case at Align level only. same for the shrink. You can also check here that range is provided instead of checking it below.

carywreams replied:
done

mechatroner commented on autoload/rainbow_csv.vim:
```diff
@@ -835,25 +834,35 @@ func! rainbow_csv#csv_align()
     endif
-    let lastLineNo = line("$")
+
+    if (l:firstline == 1 && l:lastline == line("$") && (&ft == 'markdown' || &ft == 'rmd'))
```
Please move this check higher, see my other comment.

carywreams replied:
done

mechatroner commented on autoload/rainbow_csv.vim:
```diff
@@ -820,12 +818,13 @@ func! rainbow_csv#align_field(field, is_first_line, max_field_components_lens, i
-
-func! rainbow_csv#csv_align()
+func! rainbow_csv#csv_align() range
+    let l:firstline = a:firstline
```
Please don't use `l:` prefix for consistency with other code.

carywreams replied:
updated

mechatroner commented on autoload/rainbow_csv.vim:
```diff
@@ -883,12 +892,13 @@ func! rainbow_csv#csv_align()
     endfor
     if !has_edit
-        echoerr "File is already aligned"
+        echoerr "Range is already aligned"
```
Please keep the original comment since it is a more common use case, and seeing the range might be confusing.

carywreams replied:
done

mechatroner commented on autoload/rainbow_csv.vim:
```diff
@@ -883,12 +892,13 @@ func! rainbow_csv#csv_align()
 
-
-func! rainbow_csv#csv_shrink()
+func! rainbow_csv#csv_shrink() range
```
All my comments regarding the align function also apply to Shrink function.

carywreams replied:
done.
moved the show_progress_bar setting to location similar to that found in _align, so that the exception processing block for md/rmd files could remain in a single location, as in _align.

mechatroner reviewed (changes requested):
Thank you for sending this PR!
There are some adjustments that have to be made before I can merge it.

carywreams committed:
w/r/t #51, addressing PR #52 comments and concerns

carywreams committed:
w/r/t #51, addressing PR #52; restores whitespace to README.md

carywreams committed:
w/r/t #51, addressing PR #52; restores whitespace to autoload/.vim and doc/.txt

carywreams commented on README.md:
done

mechatroner replied:
Two whitespaces at the end create a linebreak in GFM (Github-Flavored Markdown). There are multiple competing markdown dialects actually: https://en.wikipedia.org/wiki/Markdown but since this repo is hosted on github we should obviously follow the github spec.

carywreams commented on README.md:
```diff
@@ -3,73 +3,74 @@
+|csv_pipe        |&#124; (pipe)   |            |                                                      |
+|rfc_csv         |, (comma)       |            |Same as "csv" but allows multiline fields             |
+|rfc_semicolon   |; (semicolon)   |            |Same as "csv_semicolon" but allows multiline fields   |
+|markdown        |&#124; (pipe)   |.md         |align and shrink functions only operate on ranges     |
```
removed

carywreams commented on README.md:
```diff
@@ -105,65 +106,67 @@ Disable rainbow columns highlighting for the current file.
+Align CSV columns with whitespaces.
+Don't run this command if you treat leading and trailing whitespaces in fields as part of the data.
+You can edit aligned CSV file in Vim column-edit mode (Ctrl+v).
+In markdown and r-markdown files, this function requires a range of addresses or a visual range selection.
```
updated

mechatroner commented on autoload/rainbow_csv.vim:
```diff
@@ -821,12 +821,25 @@ func! rainbow_csv#align_field(field, is_first_line, max_field_components_lens, i
     let show_progress_bar = wordcount()['bytes'] > 200000
     let [delim, policy, comment_prefix] = rainbow_csv#get_current_dialect()
+    if (&ft == 'markdown' || &ft == 'rmd')
+        if (first_line == 1 && last_line == line("$"))
```
Actually, it is fine to keep your implementation if you prefer, since this is just a warning which won't be able to produce a serious bug.

mechatroner commented on autoload/rainbow_csv.vim:
```diff
@@ -821,12 +821,25 @@ func! rainbow_csv#align_field(field, is_first_line, max_field_components_lens, i
     " If there are lot of numeric columns this can additionally increase runtime by another 50% or more.
     let show_progress_bar = wordcount()['bytes'] > 200000
     let [delim, policy, comment_prefix] = rainbow_csv#get_current_dialect()
+    if (&ft == 'markdown' || &ft == 'rmd')
```
Actually, this might not be a big deal, I can fix it myself, later, so you can leave it as is if you want. "monocolumn" is just a fall back which means that no better dialect was found.

carywreams replied:
after reading through the rest of the file, wondering if &ft is correct choice over &syntax. Input?

carywreams committed:
round II, PR #52
+   is_last_line returned to treatment as bool
+   range-provided checks converted to use <range>
+   predicates exception for markdown/rmd files on current dialect =
    monocolumn for both shrink and align functions
+   adds maps of exception filetypes; used in both shink and align
    functions; other variations of markdown may be supported in future

mechatroner commented:
Thank you for your adjustments! I will take it from here, would still need to test it and maybe make some additional changes. I currently have some other competing priorities but when I have some free time on my hands I will finalize the integration. Stay tuned!

asperpharma reviewed (commented):
s

asperpharma reviewed (approved):
xs

asperpharma reviewed (approved):
v
